File: JavaScript/416-Partition Equal Subset Sum.js
/**
 * @param {number[]} nums
 * @return {boolean}
 */
var canPartition = function(nums) {
    let sum = nums.reduce((a,b)=>a+b, 0);
    if (sum%2){
        return false;
    }
    
    let subsetSum = sum/2;
    
    let dp = new Array(subsetSum+1).fill(false);
    dp[0] = true;
    for (let num of nums){
        for (let j = subsetSum; j>=num; j--){
                dp[j] |= dp[j-num];
        }    
    }
    return dp[dp.length-1];
};

File: C++/416-Partition Equal Subset Sum.cpp
class Solution {
public:
    bool canPartition(vector<int>& nums) {
        int tot = accumulate(begin(nums), end(nums), 0);
        if (tot%2) return false;

        int target = tot/2;

        vector<bool> dp(target+1, false);
        dp[0] = true;

        for (int num:nums) {
            for (int t = target; t > 0; t--) {
                if (t - num < 0) break;
                dp[t] = dp[t] || dp[t-num];
            }
        }

        return dp[target];
    }
};